File: packages/cortex-dart/test/resources/users_test.dart
import 'dart:convert';

import 'package:nfinitmonkeys_cortex_sdk/cortex_sdk.dart';
import 'package:http/http.dart' as http;
import 'package:test/test.dart';

import '../helpers.dart';

Map<String, dynamic> sampleUser({
  String id = 'user-123',
  String email = 'alice@example.com',
}) =>
    {
      'id': id,
      'email': email,
      'name': 'Alice',
      'role': 'user',
      'status': 'active',
      'createdAt': '2024-01-01T00:00:00Z',
    };

void main() {
  group('UsersResource', () {
    late CortexClient client;

    tearDown(() => client.close());

    test('list returns users', () async {
      client = CortexClient(
        apiKey: 'sk-cortex-test-key-1234',
        httpClient: mockClient(200, {
          'data': [
            sampleUser(id: 'user-1', email: 'alice@example.com'),
            sampleUser(id: 'user-2', email: 'bob@example.com'),
          ],
        }),
      );

      final users = await client.users.list();
      expect(users, hasLength(2));
      expect(users.first.id, 'user-1');
      expect(users.last.email, 'bob@example.com');
    });

    test('list handles empty response', () async {
      client = CortexClient(
        apiKey: 'sk-cortex-test-key-1234',
        httpClient: mockClient(200, {'data': <dynamic>[]}),
      );

      final users = await client.users.list();
      expect(users, isEmpty);
    });

    test('pendingCount returns count', () async {
      client = CortexClient(
        apiKey: 'sk-cortex-test-key-1234',
        httpClient: mockClient(200, {'count': 5}),
      );

      final result = await client.users.pendingCount();
      expect(result.count, 5);
    });

    test('update sends PATCH request', () async {
      final requests = <http.Request>[];
      client = CortexClient(
        apiKey: 'sk-cortex-test-key-1234',
        httpClient: recordingMockClient(200, sampleUser(), requests),
      );

      final user = await client.users.update(
        id: 'user-123',
        name: 'Updated Name',
        role: 'admin',
      );

      expect(user.id, 'user-123');
      expect(requests.first.method, 'PATCH');

      final body = jsonDecode(requests.first.body) as Map<String, dynamic>;
      expect(body['name'], 'Updated Name');
      expect(body['role'], 'admin');
    });

    test('update validates empty id', () async {
      client = CortexClient(
        apiKey: 'sk-cortex-test-key-1234',
        httpClient: mockClient(200, {}),
      );

      expect(
        () => client.users.update(id: ''),
        throwsA(isA<CortexValidationException>()),
      );
    });

    test('delete sends DELETE request', () async {
      final requests = <http.Request>[];
      client = CortexClient(
        apiKey: 'sk-cortex-test-key-1234',
        httpClient: recordingMockClient(200, {}, requests),
      );

      await client.users.delete('user-123');

      expect(requests.first.method, 'DELETE');
      expect(requests.first.url.path, contains('/admin/users/user-123'));
    });

    test('delete validates empty id', () async {
      client = CortexClient(
        apiKey: 'sk-cortex-test-key-1234',
        httpClient: mockClient(200, {}),
      );

      expect(
        () => client.users.delete(''),
        throwsA(isA<CortexValidationException>()),
      );
    });

    test('approve sends POST request', () async {
      final requests = <http.Request>[];
      client = CortexClient(
        apiKey: 'sk-cortex-test-key-1234',
        httpClient: recordingMockClient(
          200,
          sampleUser()..['status'] = 'active',
          requests,
        ),
      );

      final user = await client.users.approve('user-123');

      expect(user.id, 'user-123');
      expect(requests.first.method, 'POST');
      expect(
        requests.first.url.path,
        contains('/admin/users/user-123/approve'),
      );
    });

    test('approve validates empty id', () async {
      client = CortexClient(
        apiKey: 'sk-cortex-test-key-1234',
        httpClient: mockClient(200, {}),
      );

      expect(
        () => client.users.approve(''),
        throwsA(isA<CortexValidationException>()),
      );
    });

    test('reject sends POST request', () async {
      final requests = <http.Request>[];
      client = CortexClient(
        apiKey: 'sk-cortex-test-key-1234',
        httpClient: recordingMockClient(200, sampleUser(), requests),
      );

      final user = await client.users.reject('user-123');

      expect(user.id, 'user-123');
      expect(requests.first.method, 'POST');
      expect(
        requests.first.url.path,
        contains('/admin/users/user-123/reject'),
      );
    });

    test('resetPassword sends POST request', () async {
      final requests = <http.Request>[];
      client = CortexClient(
        apiKey: 'sk-cortex-test-key-1234',
        httpClient: recordingMockClient(
          200,
          {'temporaryPassword': 'temp-pass-123'},
          requests,
        ),
      );

      final result = await client.users.resetPassword('user-123');

      expect(result['temporaryPassword'], 'temp-pass-123');
      expect(requests.first.method, 'POST');
      expect(
        requests.first.url.path,
        contains('/admin/users/user-123/reset-password'),
      );
    });

    test('resetPassword validates empty id', () async {
      client = CortexClient(
        apiKey: 'sk-cortex-test-key-1234',
        httpClient: mockClient(200, {}),
      );

      expect(
        () => client.users.resetPassword(''),
        throwsA(isA<CortexValidationException>()),
      );
    });

    test('handles 401 errors', () async {
      client = CortexClient(
        apiKey: 'sk-cortex-test-key-1234',
        maxRetries: 0,
        httpClient: mockClient(401, 'Unauthorized'),
      );

      expect(
        () => client.users.list(),
        throwsA(isA<CortexAuthenticationException>()),
      );
    });
  });
}

File: packages/cortex-js/tests/resources/users.test.ts
import { describe, it, expect } from 'vitest';
import { ValidationError } from '../../src/errors.js';
import { mockFetch, createClient } from '../helpers.js';

describe('users', () => {
  describe('list', () => {
    it('lists users', async () => {
      const mockResponse = {
        users: [{ id: 'u-1', email: 'test@example.com', createdAt: '2024-01-01', updatedAt: '2024-01-01' }],
      };
      const fetch = mockFetch(mockResponse);
      const client = createClient(fetch);

      const result = await client.users.list();

      expect(result.users).toHaveLength(1);
      expect(result.users[0]!.email).toBe('test@example.com');

      const [url, opts] = fetch.mock.calls[0] as [string, RequestInit];
      expect(url).toBe('https://admin.test.com/admin/users');
      expect(opts.method).toBe('GET');
    });
  });

  describe('pendingCount', () => {
    it('gets pending approval count', async () => {
      const fetch = mockFetch({ count: 5 });
      const client = createClient(fetch);

      const result = await client.users.pendingCount();

      expect(result.count).toBe(5);

      const [url, opts] = fetch.mock.calls[0] as [string, RequestInit];
      expect(url).toBe('https://admin.test.com/admin/users/pending-count');
      expect(opts.method).toBe('GET');
    });
  });

  describe('update', () => {
    it('updates a user', async () => {
      const fetch = mockFetch({ id: 'u-1', email: 'test@example.com', role: 'admin', createdAt: '2024-01-01', updatedAt: '2024-01-02' });
      const client = createClient(fetch);

      const result = await client.users.update('u-1', { role: 'admin' });

      expect(result.role).toBe('admin');

      const [url, opts] = fetch.mock.calls[0] as [string, RequestInit];
      expect(url).toBe('https://admin.test.com/admin/users/u-1');
      expect(opts.method).toBe('PATCH');
    });

    it('throws ValidationError when id is missing', async () => {
      const client = createClient(mockFetch({}));
      await expect(client.users.update('', { role: 'admin' })).rejects.toThrow(ValidationError);
    });
  });

  describe('delete', () => {
    it('deletes a user', async () => {
      const fetch = mockFetch({ id: 'u-1', deleted: true });
      const client = createClient(fetch);

      const result = await client.users.delete('u-1');

      expect(result.deleted).toBe(true);

      const [url, opts] = fetch.mock.calls[0] as [string, RequestInit];
      expect(url).toBe('https://admin.test.com/admin/users/u-1');
      expect(opts.method).toBe('DELETE');
    });

    it('throws ValidationError when id is missing', async () => {
      const client = createClient(mockFetch({}));
      await expect(client.users.delete('')).rejects.toThrow(ValidationError);
    });
  });

  describe('approve', () => {
    it('approves a pending user', async () => {
      const fetch = mockFetch({ id: 'u-1', status: 'approved' });
      const client = createClient(fetch);

      const result = await client.users.approve('u-1');

      expect(result.status).toBe('approved');

      const [url, opts] = fetch.mock.calls[0] as [string, RequestInit];
      expect(url).toBe('https://admin.test.com/admin/users/u-1/approve');
      expect(opts.method).toBe('POST');
    });

    it('throws ValidationError when id is missing', async () => {
      const client = createClient(mockFetch({}));
      await expect(client.users.approve('')).rejects.toThrow(ValidationError);
    });
  });

  describe('reject', () => {
    it('rejects a pending user', async () => {
      const fetch = mockFetch({ id: 'u-1', status: 'rejected' });
      const client = createClient(fetch);

      const result = await client.users.reject('u-1');

      expect(result.status).toBe('rejected');

      const [url, opts] = fetch.mock.calls[0] as [string, RequestInit];
      expect(url).toBe('https://admin.test.com/admin/users/u-1/reject');
      expect(opts.method).toBe('POST');
    });

    it('throws ValidationError when id is missing', async () => {
      const client = createClient(mockFetch({}));
      await expect(client.users.reject('')).rejects.toThrow(ValidationError);
    });
  });

  describe('resetPassword', () => {
    it('resets a user password', async () => {
      const fetch = mockFetch({ id: 'u-1', temporaryPassword: 'tmp123' });
      const client = createClient(fetch);

      const result = await client.users.resetPassword('u-1');

      expect(result.temporaryPassword).toBe('tmp123');

      const [url, opts] = fetch.mock.calls[0] as [string, RequestInit];
      expect(url).toBe('https://admin.test.com/admin/users/u-1/reset-password');
      expect(opts.method).toBe('POST');
    });

    it('throws ValidationError when id is missing', async () => {
      const client = createClient(mockFetch({}));
      await expect(client.users.resetPassword('')).rejects.toThrow(ValidationError);
    });
  });
});
